armory/instrument/instrument.py

```diff
@@ -0,0 +1,693 @@
+        self.final = final
+        self.final_name = final_name
+        self.final_kwargs = final_kwargs or {}
+        if not isinstance(self.final_kwargs, dict):
+            raise ValueError(f"final_kwargs must be None or a dict, not {final_kwargs}")
+        self.record_final_only = record_final_only
+        self.never_measured = True
+
+    def get_arg_names(self):
+        return list(self.arg_index)
+
+    def clear(self):
+        """
+        Clear all of the current values
+        """
+        self.values = [None] * len(self.arg_index)
+        self.values_set = [False] * len(self.arg_index)
+        self.arg_batch_indices = [None] * len(self.arg_index)
+
+    def add_writer(self, writer):
+        """
+        Add a writer to emit outputs to.
+        """
+        if writer not in self.writers:
+            self.writers.append(writer)
+
+    def set(self, name, value, batch):
+        """
+        Set the value for the arg name
+        """
+        if name not in self.arg_index:
+            raise ValueError(f"{name} not a valid arg name from {self.arg_index}")
+        i = self.arg_index[name]
+        self.values[i] = value
+        self.values_set[i] = True
+        self.arg_batch_indices[i] = batch
+        if self.auto_measure and self.is_ready():
+            self.measure()
+
+    def is_ready(self, raise_error=False):
+        """
+        Return True if all values have been set and batch numbers match
+            if raise_error is True, raise ValueError instead of returning False
+        """
+        if not all(self.values_set):
+            if raise_error:
+                raise ValueError(f"Not all values have been set: {self.values_set}")
+            return False
+        if any(self.arg_batch_indices[0] != batch for batch in self.arg_batch_indices):
+            if raise_error:
+                raise ValueError(
+                    "Batch numbers are mismatched: {self.arg_batch_indices}"
+                )
+            return False
+        return True
+
+    def measure(self, clear_values=True):
+        self.is_ready(raise_error=True)
+        result = self.metric(*self.values, **self.metric_kwargs)
+        record = (self.name, self.arg_batch_indices[0], result)
+        # Assume metric is sample-wise, but computed on a batch of samples
+        try:
+            self._results.extend(result)
+        except TypeError:
+            self._results.append(result)
+        if not self.record_final_only:
+            for writer in self.writers:
+                writer.write(record)
+        if clear_values:
+            self.clear()
+        self.never_measured = False
+
+    def finalize(self):
+        """
+        Primarily intended for metrics of metrics, like mean or stdev of results
+        """
+        if self.never_measured:
+            unset = [arg for arg, i in self.arg_index.items() if not self.values_set[i]]
+            log.warning(
+                f"Meter '{self.name}' was never measured. "
```

Comment:
Need to look more closely at how/where the flag is propagated w.r.t. when/where the Meters are created for a scenario but: it seems odd that if you set `--skip-attack` you get a bunch of warnings saying that `Meter adversarial_*` was never measured. If `--skip-attack` is set, I'd think those meters shouldn't be created in the first place

Reply:
Update: I would propose, in the `add_tasks_wrt_benign_predictions()` method, changing

```
if self.task is not None:
```

to 

```
if self.task is not None and self.include_adversarial:
```

This way, we don't create meters that we know will never be used. The lone exception is the perturbation meter, for which I still see a warning that it's never used

Reply:
similarly, in the `__init__` of the `MetricsLogger` class, you could change

```
if perturbation is not None:
```

to 

```
if perturbation is not None and self.include_adversarial:
```

to avoid creating a perturbation meter that we know won't be used (and thus get rid of the warning)

Reply:
Updated.